src/ui/control/geolocate_control.js

```diff
@@ -107,58 +247,177 @@ class GeolocateControl extends Evented {
             this._container);
         this._geolocateButton.type = 'button';
         this._geolocateButton.setAttribute('aria-label', 'Geolocate');
-        if (this.options.watchPosition) this._geolocateButton.setAttribute('aria-pressed', false);
+
+        if (this.options.trackUserLocation) {
+            this._geolocateButton.setAttribute('aria-pressed', false);
+            this._watchState = 'OFF';
+        }
+
+        // when showUserLocation is enabled, keep the Geolocate button disabled until the device location marker is setup on the map
+        if (this.options.showUserLocation) {
+            if (this.options.trackUserLocation) this._watchState = 'INITILIZE';
+
+            this._dotElement = DOM.create('div', 'mapboxgl-user-location-dot');
+
+            // until https://github.com/mapbox/mapbox-gl-js/issues/2900, the offset is used to ensure the Marker is centered at the
+            // user's location, half of the .mapboxgl-user-location-dot width (including border).
+            this._userLocationDotMarker = new Marker(this._dotElement, { offset: [-10, -10] });
+
+            if (this.options.trackUserLocation) this._watchState = 'OFF';
+        }
+
         this._geolocateButton.addEventListener('click',
             this._onClickGeolocate.bind(this));
+
+        // when the camera is changed (and it's not as a result of the Geolocation Control) change
+        // the watch mode to background watch, so that the marker is updated but not the camera.
+        if (this.options.trackUserLocation) {
+            this._map.on('movestart', (event) => {
+                if (!event.geolocateSource) {
+                    if (this._watchState === 'ACTIVE_LOCK') {
```

Comment:
Nit: change the nested `if`s to `if(!event.geolocateSource && this._watchState == 'ACTIVE_LOCK
'`)